File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int n = s.length();

        int left = 0;
        int max = 0;
        Set<Character> set = new HashSet<>();

        for(int right = 0;right<n;right++){
            if(set.contains(s.charAt(right))){
                while(set.contains(s.charAt(right))){
                    set.remove(s.charAt(left));
                    left++;
                }
            }
            set.add(s.charAt(right));
            int length = right - left +1;
            max = Math.max(length, max);
        }
        return max;
        
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
var lengthOfLongestSubstring = function (s) {
  let left = 0;
  let right = 0;
  let max = 0;
  let set = new Set();

  while (right < s.length) {
    if (!set.has(s[right])) {
      set.add(s[right]);
      max = Math.max(max, right - left + 1);
      right++;
    } else {
      set.delete(s[left]);
      left++;
    }
  }

  return max;
};
